### PR TITLE
Support for grouping

### PR DESCRIPTION
**Pull request by luke-segars**
This is a pretty big one so I've broken it up into three revisions for review. Happy to collapse them later.

A general overview of the design is here: https://docs.google.com/document/d/1aIjHNzWSw3Y1Zm3ndEjmkV4yjL2hUlxaQJo_aGI_dgE/edit#. Note that I wrote this up beforehand so there may be some slight discrepancies.

This PR introduces the UserGroup class, which represents groups of users that can communicate together (via group whisper) and share resources (such as experience). A few things to note:

   - This introduces support for two opcodes: **0x2E** (called GroupRequest) and **0x2F** (called GroupToggle).
   - Group joining / leaving interactions have been kept as similar as I could to what USDA does, with the exception of some messaging. Certain messages are changed to strings that are more sensible ("You've joined a group" instead of "<your name> has joined a group", for example).
   - This introduces a new constant called **GROUP_SHARING_DISTANCE** that defines the maximum Manhattan distance between a user and an event (death of a monster) where the user will be eligible to share in the spoils.
   - This introduces a new experience function called **ShareExperience** that calls a distribution function if the user is grouped or falls back to **GiveExperience** if not. The intent is that this function would be called whenever experience should be shared if a user is grouped (hunting, etc), but GiveExperience would be used whenever experience shouldn't be shared despite group status (solo quests, etc).
   - Whisper handling now branches depending on the name of the target. If the name == "!!" (magic sequence from USDA) then it's a group whisper, otherwise it's a direct whisper. I abstracted a bit of pre-existing logic to make sure that muting rules are shared between group and direct whispers.
   - Added /group, /ungroup, and /exp commands to make all of this stuff easier to test.
   - Resource distribution functions are intended to be easy to extend and modify (though they're not scripted, so they still need to be compiled in). Could be a good candidate for scripting in the future (maybe?).

Thoughts welcome; thanks! Probably going to make another cleanup pass on this but wanted to get some feedback on the general design and whether there were any obvious missing pieces as well.

### CODE REVIEW

**Comment by luke-segars** [comment]
Also, worth pointing out a few (I think small) issues that I know about. I might try to fix them here or fix them in a future PR.

   - Some groups don't seem to be released for garbage collection, so there's potentially an opportunity for memory leaks. I haven't been able to tell what the risk size is, though I have confirmed that some groups are cleaned up.
   - At some point there was a key users could press to highlight everyone who was in their group. I haven't started on that but I couldn't find what the key was supposed to be (the original key, H, seems to have been repurposed). It also feels like a pretty easy follow-up later rather than continuing to bloat this PR.

**Comment by luke-segars** [comment]
Last but not least, this PR is addressing this bug: https://hybrasyl.atlassian.net/browse/SERVER-91.

And now I realize that that bug calls out J as the mystery highlight character :) might still follow up with that later.

File: hybrasyl/World.cs [hunk 1]
```diff
@@ -1963,6 +2018,111 @@ private void PacketHandler_0x2D_PlayerInfo(Object obj, ClientPacket packet)
...
+                // request. We need to add them to the original user's group. Note that in this
+                // case the partner sent the original invitation.
+                case 0x03:
+                    Logger.Info("Invitation accepted. Grouping.");
+                    partner.InviteToGroup(user);
+                    break;
+                // This shouldn't happen but we should log it and fix it if it does.
+                default:
+                    Logger.Info("Unknown GroupRequest stage. No action taken.");
+                    break;
```
Comment (baughj) [inline hybrasyl/World.cs]: All of these Info level logs should be going to debug instead, I think, in order to not clutter up the logfile. The Unknown stage should also be at Error level.

**Comment by baughj** [comment]
:+1: - just think we need the logging level changes and we're good. How extensively have you tested this?

**Comment by luke-segars** [comment]
Thanks for the review! I don't have any proper tests written for this stuff yet. All of the testing has been manual spot checking but I've tried to break it and (after succeeding many times) haven't been able to find issues than those mentioned above. I'll make another pass later this week and confirm here with the results.

It's also a non-trivial # of changes so let me think for a bit on what I can do on the testing front. It'd make me more comfortable with it.

**Comment by baughj** [comment]
I think in most cases for now the testing will just have to be done on staging with actual people.

**Comment by baughj** [comment]
@luke-segars were you able to test this at all? The code looks reasonable to me, I think we should just merge and test in staging (in the absence of unit testing / etc, this is basically how our process works).

**Comment by luke-segars** [comment]
Unfortunately I've been traveling a ton this week and haven't had a chance.
I have a PR to change logging levels and will send that over soon. From a
testing perspective I think we're OK though.

**Comment by luke-segars** [comment]
Alrighty I've changed log levels and did a bit of spot testing. I think we're ready to merge if all looks good to you @baughj! Thanks @kojasou for your help dissecting packets for this.

@baughj can we also close the JIRA bug for this?

**Comment by baughj** [comment]
Thanks!

**Pull request merged**